drivers/kscan/kscan_cst816s.c

```diff
@@ -0,0 +1,333 @@
+#define CST816S_REG_BPC1L               0xB3
+#define CST816S_REG_POWER_MODE          0xA5
+#define CST816S_REG_CHIP_ID             0xA7
+#define CST816S_REG_PROJ_ID             0xA8
+#define CST816S_REG_FW_VERSION          0xA9
+#define CST816S_REG_MOTION_MASK         0xEC
+#define CST816S_REG_IRQ_PULSE_WIDTH     0xED
+#define CST816S_REG_NOR_SCAN_PER        0xEE
+#define CST816S_REG_MOTION_S1_ANGLE     0xEF
+#define CST816S_REG_LP_SCAN_RAW1H       0xF0
+#define CST816S_REG_LP_SCAN_RAW1L       0xF1
+#define CST816S_REG_LP_SCAN_RAW2H       0xF2
+#define CST816S_REG_LP_SCAN_RAW2L       0xF3
+#define CST816S_REG_LP_AUTO_WAKEUP_TIME 0xF4
+#define CST816S_REG_LP_SCAN_TH          0xF5
+#define CST816S_REG_LP_SCAN_WIN         0xF6
+#define CST816S_REG_LP_SCAN_FREQ        0xF7
+#define CST816S_REG_LP_SCAN_I_DAC       0xF8
+#define CST816S_REG_AUTOSLEEP_TIME      0xF9
+#define CST816S_REG_IRQ_CTL             0xFA
+#define CST816S_REG_DEBOUNCE_TIME       0xFB
+#define CST816S_REG_LONG_PRESS_TIME     0xFC
+#define CST816S_REG_IOCTL               0xFD
+#define CST816S_REG_DIS_AUTO_SLEEP      0xFE
+
+#define CST816S_MOTION_EN_CON_LR      BIT(2)
+#define CST816S_MOTION_EN_CON_UR      BIT(1)
+#define CST816S_MOTION_EN_DCLICK      BIT(0)
+
+#define CST816S_IRQ_EN_TEST           BIT(7)
+#define CST816S_IRQ_EN_TOUCH          BIT(6)
+#define CST816S_IRQ_EN_CHANGE         BIT(5)
+#define CST816S_IRQ_EN_MOTION         BIT(4)
+#define CST816S_IRQ_ONCE_WLP          BIT(0)
+
+#define CST816S_IOCTL_SOFT_RTS        BIT(2)
+#define CST816S_IOCTL_IIC_OD          BIT(1)
+#define CST816S_IOCTL_EN_1V8          BIT(0)
+
+#define CST816S_POWER_MODE_SLEEP          (0x03)
+#define CST816S_POWER_MODE_EXPERIMENTAL   (0x05)
+
+#define EVENT_PRESS_DOWN	0x00U
+#define EVENT_LIFT_UP		0x01U
+#define EVENT_CONTACT		0x02U
+#define EVENT_NONE		0x03U
+
+
+/** cst816s configuration (DT). */
+struct cst816s_config {
+	struct i2c_dt_spec i2c;
+	const struct gpio_dt_spec rst_gpio;
+#ifdef CONFIG_KSCAN_CST816S_INTERRUPT
+	const struct gpio_dt_spec int_gpio;
+#endif
+};
+
+/** cst816s data. */
+struct cst816s_data {
+	/** Device pointer. */
+	const struct device *dev;
+	/** KSCAN Callback. */
+	kscan_callback_t callback;
+	/** Work queue (for deferred read). */
+	struct k_work work;
+
+#ifdef CONFIG_KSCAN_CST816S_INTERRUPT
+	/** Interrupt GPIO callback. */
+	struct gpio_callback int_gpio_cb;
+#else
+	/** Timer (polling mode). */
+	struct k_timer timer;
+#endif
+};
+
+static int cst816s_process(const struct device *dev)
+{
+	const struct cst816s_config *cfg = dev->config;
+	struct cst816s_data *data = dev->data;
+
+	int r;
+	uint8_t event;
+	uint16_t row, col;
+	bool pressed;
+	uint8_t buf[9];
+
+	r = i2c_burst_read_dt(&cfg->i2c,
+				CST816S_REG_DATA, buf, 9);
+	if (r < 0) {
+		LOG_ERR("Could not read data");
+		return r;
+	}
+
+	col = ((buf[CST816S_REG_XPOS_H] & 0x0f) << 8) | buf[CST816S_REG_XPOS_L];
+	row = ((buf[CST816S_REG_YPOS_H] & 0x0f) << 8) | buf[CST816S_REG_YPOS_L];
```

Comment:
`<sys/byteorder.h>` can be used

Reply:
not solved

Reply:
col = ((buf[3] & 0x0f) << 8) | buf[4];
row = ((buf[5] & 0x0f) << 8) | buf[6];
this buf are not aligned, so maybe not suitable